keps/sig-api-machinery/0006-apply.md

```diff
@@ -443,6 +445,101 @@ y,Initialized,PodScheduled,Ready,/var/run/secrets/kubernetes.io/serviceaccount]}
+  sense to add it to the dictionary.
+- Since the base64 strings of low ints are shorter, they should be used for
+  more frequently used strings in fieldsets.
+
+Fine tuning the string table based on what strings are most frequent is
+something we can iterate on. If the initial string table size is small, ordering
+by frequency might not matter that much, but if/when the table gets larger,
+ensuring that frequently used strings are assigned low numbers in the table will
+become increasingly valuable.
+
+There are plenty of potential future optimizations, like gathering fieldset data
+from real-world Kubernetes clusters at runtime, and using it as training data to
+improve string tables.
+
+##### Versioning String Tables
+
+Each string table has a version, starting at 1:
+
+```
+String table:
+{<versionNumber>:[...
+```
+
+Fieldsets serialized using string tables contain the version they were serialized with:
+
+```
+Field set:
+[<versionNumber>, ...
+```
+
+Since a fieldset can only be deserialized using the versioned string table
+it was serialized with, each Kubernetes version must be able to deserialize
+fieldsets that were serialized by:
+
+- All of the prior Kubernetes minor versions.
+- The next Kubernetes minor version, so that N-1 rollback is safe
+
+We can ensure this by:
+
+- Keeping all string table versions around forever
+- Introducing a string table version to Kubernetes one minor version before
+  serializing data using that string table version
+
+Alternative considered: We could instead persist string table versions in etcd--
+each time the kube-apiserver starts, it writes to etcd any string table versions
+it knows about that are not alreeay stored in etcd. Then, if the kube-apiserver
+ever needs to deserialize a fieldset with a string table version it is not aware
+of, it tries to read the needed string table version from etcd. This would allow
+for both HA upgrades and rollbacks. But it is complicated, and since we don't
+want or need to add new the string table versions often, it is unnecessary.
+
+Logically, the information Kubernetes needs about the string tables are:
+
+- The versions of string table it is aware of
+- Which version of the string table it should use when serializing
+
+```
+// Current Kubernetes version: 1.18
+
+const serializeVersion = 1
+
+const stringTableVersions = {
+  1:{[...], introducedVersion: "1.17"}, // may be used to serialize in kubernetes 1.18
+  2:{[...], introducedVersion: "1.18"}, // may be used to serialize in kubernetes 1.19
+  ...
+```
+
+Automated checks we will add to ensure the string tables are versioned correctly:
+
+- Check that the "introducedVersion" of the current "serailizeVersion is less
+  than the current kubernetes (for n-1 rollback support).
+- Check that that string table where the "introducedVersion" is less than the
+  curreht kubernetes version are not modified. E.g.  check the the
+  hashe/checksum of these string tables.
```

Comment:
s/hashe/hash/

We should store the hash in smd with the string tables to help people keep them immutable.